### PR TITLE
Migrate Place

### PR DESCRIPTION
**Pull request by cbaines**
Migrate the Place format, such that it is rendered through the Publishing API.

Do not merge yet, as the merging/release of this needs to be coordinated with Frontend.

### CODE REVIEW

**Comment by davidbasalla** [comment]
This should now add a `document_type` method to the Presenter, see https://github.com/alphagov/publisher/pull/581

File: app/presenters/formats/place_presenter.rb [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      'place'
+    end
+
+    def details
+      details = { external_related_links: external_related_links }
+
+      details = details.merge(govspeak_fields)
+
+      if @edition.place_type
+        details = details.merge(place_type: @edition.place_type)
```
Comment (whoojemaflip) [inline app/presenters/formats/place_presenter.rb]: there's an attribute reader for the `@edition` variable, so we can lose the snail.

File: test/unit/presenters/formats/place_presenter_test.rb [hunk 2]
```diff
@@ -0,0 +1,65 @@
+require 'test_helper'
```
Comment (whoojemaflip) [inline test/unit/presenters/formats/place_presenter_test.rb]: could this be in the same commit as the class under test?

**Comment by whoojemaflip** [comment]
This could be squashed

File: app/presenters/formats/place_presenter.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    end
+
+    def document_type
+      'place'
+    end
+
+    def details
+      details = { external_related_links: external_related_links }
+
+      details = details.merge(govspeak_fields)
```
Comment (whoojemaflip) [inline app/presenters/formats/place_presenter.rb]: It's a little hard to follow what keys this adds to the details hash, since the implementation is scattered through the whole class.  Could it be structured so that the keys are inline in this method to make it more obvious what it returns?

File: app/presenters/formats/place_presenter.rb [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+      if edition.place_type
+        details = details.merge(place_type: edition.place_type)
```
Comment (whoojemaflip) [inline app/presenters/formats/place_presenter.rb]: The `merge` method creates a new hash, copies the combined contents into it, then we're assigning that new hash to the same local variable. 

Seems it might make more sense to just modify the existing hash in-place.

`details[:place_type] = edition.place_type`

File: app/presenters/formats/place_presenter.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    end
+
+    def govspeak_fields
+      %i(introduction more_information need_to_know).inject({}) do |details, field|
+        next details if edition[field].blank?
+
+        details[field] = [
+          {
+            content_type: "text/govspeak",
+            content: edition[field]
```
Comment (whoojemaflip) [inline app/presenters/formats/place_presenter.rb]: I'm a bit confused by the `[]` operator here - is it common to access properties of the `Edition` model using the hash accessor?

Comment (whoojemaflip) [inline app/presenters/formats/place_presenter.rb]: If so, it's a neat trick for working with the dynamic property thing.

File: app/presenters/formats/place_presenter.rb (hunk 1 above)
Comment (cbaines) [inline app/presenters/formats/place_presenter.rb]: I wrote this, but I think @brenetic improved it as you describe :)

File: test/unit/presenters/formats/place_presenter_test.rb (hunk 2 above)
Comment (cbaines) [inline test/unit/presenters/formats/place_presenter_test.rb]: Done.

File: test/unit/presenters/formats/place_presenter_test.rb [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+  def artefact
+    @_artefact ||= FactoryGirl.create(:artefact, kind: "place", slug: "find-food")
+  end
+
+  def result
+    subject.render_for_publishing_api
+  end
+
+  should "be valid against schema" do
+    puts result
```
Comment (cbaines) [inline test/unit/presenters/formats/place_presenter_test.rb]: Yep, removed :)

File: app/presenters/formats/place_presenter.rb (hunk 3 above)
Comment (cbaines) [inline app/presenters/formats/place_presenter.rb]: I've changed it to hopefully make it easier to read, let me know what you think.

File: app/presenters/formats/place_presenter.rb (hunk 5 above)
Comment (cbaines) [inline app/presenters/formats/place_presenter.rb]: I believe this is a Mongoid thing.

**Comment by whoojemaflip** [comment]
WFM :sheep: :it:

**Comment by davidbasalla** [comment]
We've previously added a test to check that `#migrated?` returns true for each migrated format here `https://github.com/alphagov/publisher/blob/master/test/unit/edition_test.rb#L75-L110)` so might be good to add for consistency?

**Comment by davidbasalla** [comment]
Same here https://github.com/alphagov/publisher/blob/cefe2eecde3f6b788638323825cd9d2821804904/test/unit/services/edition_presenter_factory_test.rb

**Comment by cbaines** [comment]
> We've previously added a test to check that #migrated? returns true for each migrated format here so might be good to add for consistency?

Yep, I've done this now :)

**Comment by davidbasalla** [comment]
Looks good, merge when ready! 👍
